### PR TITLE
MM-32225 Cloud Plugin Enhancements - Installation import

### PR DESCRIPTION
**Pull request by IndushaS**
#### Summary

Add functionality to the cloud plugin for importing installations created from other sources to be managed by the cloud plugin.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-32225

#### Release Note

```release-note
Add installation import command
```

### CODE REVIEW

File: server/command.go [hunk 1]
```diff
@@ -125,7 +130,7 @@ func (p *Plugin) ExecuteCommand(c *plugin.Context, args *model.CommandArgs) (*mo
 }
 
 func (p *Plugin) runInfoCommand(args []string, extra *model.CommandArgs) (*model.CommandResponse, bool, error) {
-	resp := fmt.Sprintf("Mattermost Cloud plugin version: %s, "+
+	resp := fmt.Sprintf("Hello Mattermost Cloud plugin version: %s, "+
```
Comment (gabrieljackson) [inline server/command.go:133]:
```suggestion
	resp := fmt.Sprintf("Mattermost Cloud plugin version: %s, "+
```
Leftover debug text? :)

File: server/command_import.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		return nil, false, errors.Wrap(err, "failed to get installation by DNS")
+	}
+	if cloudInstall == nil {
+		return nil, true, errors.New("no installation for the DNS provided")
+	}
+	if cloudInstall.OwnerID != extra.UserId {
+		cloudInstall.OwnerID = extra.UserId
+	}
+
+	p.cloudClient.UpdateInstallation(cloudInstall.ID, &cloud.PatchInstallationRequest{
```
Comment (gabrieljackson) [inline server/command_import.go:35]: This call is updating the installation in the provisioner and returning an error. You aren't capturing the error here, so we should do that and check to make sure the error is `nil`.

Also, we only need to run this when `cloudInstall.OwnerID != extra.UserId` which is checked on line 31. We should probably move this inside that `if` statement.

**Review by gabrieljackson — CHANGES_REQUESTED** [review]
This is great! Just a few requests for improvements.

File: server/command_import.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	}
+	name := splitDNS[0]
+
+	cloudInstall, err := p.cloudClient.GetInstallationByDNS(installationDNS, nil)
+	if err != nil {
+		return nil, false, errors.Wrap(err, "failed to get installation by DNS")
+	}
+	if cloudInstall == nil {
+		return nil, true, errors.New("no installation for the DNS provided")
+	}
```
Comment (gabrieljackson) [inline server/command_import.go:30]: We need to check one other thing after this. We need to make sure that the installation hasn't already been imported by someone else in the cloud plugin.

To do this, we should see if the installation ID is found in the plugin database.

File: server/command_import.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+	installs, _, err := p.getInstallations()
+	if err != nil {
+		return nil, false, err
+	}
+
+	for _, install := range installs {
+		if install.ID == cloudInstall.ID {
+			return nil, true, errors.New("Installation has already been imported to cloud plugin")
```
Comment (gabrieljackson) [inline server/command_import.go:39]: nit:
```suggestion
			return nil, true, errors.New("installation has already been imported to cloud plugin")
```

**Review by gabrieljackson — APPROVED** [review]
Looks great! Just one minor comment, but no blockers.

Please test this in the dev environment like we spoke about, but otherwise LGTM.

File: server/command_import.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		}
+	}
+
+	if cloudInstall.OwnerID != extra.UserId {
+		cloudInstall.OwnerID = extra.UserId
```
Comment (jwilander) [inline server/command_import.go:44]: This means user A could import user B's installation, right? Don't think that's a problem just wanted to make sure that's expected

**Review by jwilander — APPROVED** [review]
Looks good 👍 just one non-blocking comment

File: server/command_import.go (hunk 5 above)
Comment (gabrieljackson) [reply server/command_import.go:44]: Hey @jwilander, we actually made sure to cover this case with some logic above:

```
	for _, install := range installs {
		if install.ID == cloudInstall.ID {
			return nil, true, errors.New("Installation has already been imported to cloud plugin")
		}
	}
```

This should ensure the installation isn't already in the cloud plugin before we get here.

Comment (jwilander) [reply server/command_import.go:44]: Right, my mistake. I thought that was getting and only checking against the installations for the current user but it's all of them. Got it.

**Pull request merged**